daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -975,18 +975,16 @@ sealed abstract class StepExpression(val step: String, val pred: Option[Predicat
   }
 
   final lazy val checkAmbiguousPath: Unit = {
-    val (arraysOrOptionals, scalars) = stepElements.partition { se =>
-      se.isArray || se.isOptional
-    }
-    (arraysOrOptionals.length, scalars.length) match {
-      case (a, s) if (a > 0 && s > 0) =>
-        arraysOrOptionals.head.SDE(
-          "Path step is ambiguous. It can reference both an array or a non-array element, which is not allowed.\n" +
-            "One of the non-arrays is %s",
-          scalars.head.schemaFileLocation.toString,
-        )
-      case _ => // do nothing
-    }
+    val (arrays, nonArrays) = stepElements.partition { _.isArray }
+    schemaDefinitionWhen(
+      arrays.length > 0 && nonArrays.length > 0,
+      "Path step %s is ambiguous. It can reference both an array or a non-array element, which is not allowed.\n" +
```

Comment:
I suggest changing `or` to `and` in the error message above, even though it may cause some tests to require changes to their expected output.  When you say "both", that implies "and" rather than "or".